lib/service_provider.rb

```diff
@@ -1,105 +1,65 @@
-  DEFAULT_ENCRYPTION = 'aes256-cbc'.freeze
-
   attr_reader :issuer
 
   def initialize(issuer)
     @issuer = issuer
   end
 
   def metadata
-    ignored_methods = [:metadata, :encrypt_responses?, :encryption_opts, :issuer]
-    metadata_methods = self.class.instance_methods(false) - ignored_methods
-
-    @metadata ||= metadata_methods.inject({}) do |hash, method|
-      hash.merge(:"#{method}" => send(method))
-    end
-  end
-
-  def acs_url
-    sp_attributes['acs_url']
-  end
+    raise "#{issuer.inspect} is an invalid service provider!" if sp_attributes.empty?
 
-  def assertion_consumer_logout_service_url
-    sp_attributes['assertion_consumer_logout_service_url']
+    sp_attributes.merge(shared_attributes)
   end
 
-  def sp_initiated_login_url
-    sp_attributes['sp_initiated_login_url']
-  end
-
-  def metadata_url
-    sp_attributes['metadata_url']
-  end
-
-  def agency
-    sp_attributes['agency']
-  end
-
-  def friendly_name
-    sp_attributes['friendly_name']
-  end
-
-  def attribute_bundle
-    sp_attributes['attribute_bundle']
+  def encryption_opts
+    return nil unless encrypt_responses?
+    {
+      cert: ssl_cert,
+      block_encryption: block_encryption,
+      key_transport: 'rsa-oaep-mgf1p'
+    }
   end
 
-  def cert
-    sp_cert = sp_attributes['cert']
-    return if sp_cert.blank?
-
-    cert_dir = "#{Rails.root}/certs/sp/"
-
-    @cert ||= File.read("#{cert_dir}#{sp_cert}.crt")
-  end
+  private
 
-  def block_encryption
-    sp_attributes['block_encryption'] || DEFAULT_ENCRYPTION
+  def sp_attributes
+    @sp_attributes ||= config.sp_attributes
   end
 
-  def key_transport
-    'rsa-oaep-mgf1p'
+  def config
+    ServiceProviderConfig.new(filename: 'service_providers.yml', issuer: issuer)
   end
 
-  def encryption_opts
-    return nil unless encrypt_responses?
+  def shared_attributes
     {
-      cert: OpenSSL::X509::Certificate.new(cert),
-      block_encryption: block_encryption,
-      key_transport: key_transport
+      fingerprint: fingerprint
     }
   end
 
-  def fingerprint
-    @fingerprint ||= ::Fingerprinter.fingerprint_cert(cert)
-  end
-
-  def double_quote_xml_attribute_values
-    true
-  end
+  def ssl_cert
+    @ssl_cert ||= begin
+      sp_cert = sp_attributes[:cert]
+      return if sp_cert.blank?
 
-  private
+      cert_dir = "#{Rails.root}/certs/sp/"
 
-  def config
-    @config ||= YAML.load_file("#{Rails.root}/config/service_providers.yml")
+      cert_file = File.read("#{cert_dir}#{sp_cert}.crt")
 
-    if Figaro.env.domain_name == 'superb.legit.domain.gov'
-      @config.merge!(@config.fetch('superb.legit.domain.gov', {}))
-    else
-      @config.merge!(@config.fetch(Rails.env, {}))
+      OpenSSL::X509::Certificate.new(cert_file)
     end
-
-    @config.symbolize_keys!
   end
 
-  def sp_attributes
-    config[:valid_hosts].fetch(@issuer, {})
+  def fingerprint
+    @fingerprint ||= ::Fingerprinter.fingerprint_cert(ssl_cert)
```

Comment:
Looks like it's not needed. It was already there so I didn't change it. The tests pass without the colons so I'll remove them. Presumably at one point there was an issue with accessing without the colons, probably due to a missing `require` statement somewhere, and due to Fingerprinter being inside the `lib` directory, which doesn't automatically get required by Rails.